athenz-identity-provider-service/src/main/java/com/yahoo/vespa/hosted/athenz/instanceproviderservice/AthenzSslKeyStoreConfigurator.java

```
@@ -82,35 +124,41 @@ public void deconstruct() {
-            X509Certificate certificate = certificateClient.updateCertificate(privateKey, CERTIFICATE_EXPIRY_TIME);
-            verifyActualExpiry(certificate);
+            X509Certificate certificate = certificateClient.updateCertificate(privateKey);
+            Instant expirationTime = certificate.getNotAfter().toInstant();
+            Duration expiry = Duration.between(certificate.getNotBefore().toInstant(), expirationTime);
+            log.log(LogLevel.INFO, String.format("Got Athenz x509 certificate with expiry %s (expires %s)", expiry, expirationTime));
 
             KeyStore keyStore = KeyStore.getInstance("JKS");
             keyStore.load(null);
             keyStore.setKeyEntry(
                     CERTIFICATE_ALIAS, privateKey, CERTIFICATE_PASSWORD.toCharArray(), new Certificate[]{certificate});
+            writeKeystore(keyStore, keystoreCachePath);
             return keyStore;
-        } catch (IOException | NoSuchAlgorithmException | CertificateException | KeyStoreException e) {
+        } catch (IOException | GeneralSecurityException e) {
             throw new RuntimeException(e);
         }
     }
 
-    private static void verifyActualExpiry(X509Certificate certificate) {
-        Duration actualExpiry =
-                Duration.between(certificate.getNotBefore().toInstant(),  certificate.getNotAfter().toInstant());
-        if (CERTIFICATE_EXPIRY_TIME.compareTo(actualExpiry) > 0) {
-            log.log(LogLevel.WARNING,
-                    String.format("Expected expiry %s, got %s", CERTIFICATE_EXPIRY_TIME, actualExpiry));
+    private static void writeKeystore(KeyStore keyStore, Path keystoreCachePath) {
+        try (OutputStream out = new BufferedOutputStream(new FileOutputStream(keystoreCachePath.toFile()))) {
+            keyStore.store(out, new char[0]);
+        } catch (IOException | GeneralSecurityException e) {
+            throw new RuntimeException(e);
```

Should we fail hard if we are unable to write the keystore or just log ERROR ? config server does not rely on the written cert

---

Sounds like a good idea.